msprime/branch_stats.py

```diff
@@ -0,0 +1,229 @@
+import msprime
+
+
+def path_length(tr,x,y):
+    L = 0
+    mrca = tr.mrca(x,y)
+    for u in x,y:
+        while u != mrca:
+            L += tr.branch_length(u)
+            u = tr.parent(u)
+    return L
+
+def branch_length_diversity(ts,X,Y):
+    '''
+    Computes average pairwise diversity between a random choice from x
+    and a random choice from y.
+    '''
+    S = 0
+    for tr in ts.trees():
+        SS = 0
+        for x in X:
+            for y in Y:
+                SS += path_length(tr,x,y)
+        S += SS*tr.length
+    return S/(ts.sequence_length*len(X)*len(Y))
+
+def branch_length_Y(ts,x,y,z):
+    S = 0
+    for tr in ts.trees():
+        xy_mrca = tr.mrca(x,y)
+        xz_mrca = tr.mrca(x,z)
+        yz_mrca = tr.mrca(y,z)
+        if xy_mrca == xz_mrca:
+            #   /\
+            #  / /\
+            # x y  z
+            S += path_length(tr,x,yz_mrca)*tr.length
+        elif xy_mrca == yz_mrca:
+            #   /\
+            #  / /\
+            # y x  z
+            S += path_length(tr,x,xz_mrca)*tr.length
+        elif xz_mrca == yz_mrca:
+            #   /\
+            #  / /\
+            # z x  y
+            S += path_length(tr,x,xy_mrca)*tr.length
+    return S/ts.sequence_length
+
+def branch_length_f4(ts,A,B,C,D):
```

Comment:
This method (and those above are really just for testing, so lets move these into the test_branch_stats.py file.